### PR TITLE
ZSink.collectAll and Zstream.runAll use chunk

### PR DESCRIPTION
**Pull request by borissmidt**
Change ZSink.collectAll to return Chunk
Change Zstream.runAll to return Chunk

I think i left one todo for the ZIO.foreach that returns a list as well but it would make this PR too big.

There was on inconvenience with the current test system i only had 1 compilation error left 
but because it didn't check all test cases i had to compile many times with only a single error reported.

#3575

### CODE REVIEW

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 1]
```diff
@@ -481,10 +481,10 @@ object ZSink extends ZSinkPlatformSpecificConstructors {
     new ZSink(push) {}
 
   /**
-   * A sink that collects all of its inputs into a list.
+   * A sink that collects all of its inputs into a chunk.
    */
-  def collectAll[A]: ZSink[Any, Nothing, A, List[A]] =
-    foldLeftChunks(Chunk[A]())(_ ++ (_: Chunk[A])).map(_.toList)
+  def collectAll[A]: ZSink[Any, Nothing, A, Chunk[A]] =
+    foldLeftChunks(Chunk[A]())(_ ++ (_: Chunk[A]))
```
[the remark this answers is not in the document]
Comment (borissmidt) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala:487]: Good catch, i'll see ik there is any other place with a similar mistake.

**Comment by borissmidt** [comment]
> Taking a step back here, @borissmidt I'd like to see a benchmark of this first.
> 
> Building a list (and yes, reversing it) is very fast in Scala because the appends are O(1) and the reversal can happen with a pre-allocated buffer.
> 
> According to this: https://www.lihaoyi.com/post/BenchmarkingScalaCollections.html#construction-performance the only thing that is faster than prepending to a list is pre-allocating an array and filling it up, but we can't do that because we don't know up-front what is the size of the stream.
> 
> Another thing to consider when collecting to a chunk is the size limit: a ChunkBuilder will accumulate elements into an array, so this has a hard size limit at `2^31`. Not sure if this is a showstopper or not.

Maybe it is a better idea to change it to Seq then as the return type then you can change it on to something that would give the best performance without breaking binary compatibility? 
 
Since i can imagine that there are other better structures than an array that has to be copied every so often and a list that has a lot of indirection. Also in case of reversing it has to iterate over the whole list?

How could i best benchmark this? 

looking at the scala 2.13 source for the reverse it doesn't use a pre-allocated buffer so you need to iterate the full list.
```
  final override def reverse: List[A] = {
    var result: List[A] = Nil
    var these = this
    while (!these.isEmpty) {
      result = these.head :: result
      these = these.tail
    }
    result
  }
```

**Comment by borissmidt** [comment]
I've updated the scala-bench project to use builders and will run them overnight.
It could be used to benchmark the `chunk` collection?

https://github.com/borissmidt/scala-bench

**Comment by iravid** [comment]
@borissmidt I think it could be easier to just add another test case to `ChunkBenchmarks.scala` in this project.

**Comment by borissmidt** [comment]
[scala-bench.zip](https://github.com/zio/zio/files/4640007/scala-bench.zip)
I was just curios my self to see if the array builder is faster than the normal list add operation and indeed it is. results are the average runtime in nano seconds of iteration of a run.

**Comment by iravid** [comment]
Can you post the results here?

**Comment by iravid** [comment]
You were testing list prepend, not append, right?

**Comment by borissmidt** [comment]
Yes list prepend! the :: operation. you can view the code and run it yourself i've posted it.

**Comment by iravid** [comment]
Right. I'm looking at the results of `ListReverse` and `ArrayBuilder` which is what we'd be using in practice. At 250k elements, the `List` becomes faster.

I think this is good evidence that `ChunkBuilder` and `Chunk` are good for the `runCollect` usecase!

**Comment by borissmidt** [comment]
I think the added advantage with the chunk is that it's part of the
library. So you can improve it.

Maybe you could make it a linked list of arrays of at most 128k elements :)

**Comment by simpadjo** [comment]
ZSink is expected to be the final stage of stream processing. If a user decided to collect a huge stream into a List/Chunk, it opted-in for having abnormally large objects in its program, and will face problems anyway. 
I don't think we should spend too much of an effort optimizing for this case.

@borissmidt you can get rid of linting errors by running `sbt prepare` (see https://github.com/zio/zio/blob/master/docs/about/contributing.md)

**Comment by borissmidt** [comment]
There is still the open decision to keep the old Api and add the chunk as separate function calls.
Or replace them with chunk as i've done now.

**Comment by borissmidt** [comment]
`"size can be modified locally" at src/test/scala/zio/test/GenSpec.scala:607` fails and i don't really understand why this is related to the changes i had made.

**Comment by iravid** [comment]
@adamgfraser Can you peek at the test failure here and give us a hint?

**Comment by adamgfraser** [comment]
@iravid Yes let me take a look.

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 2]
```diff
@@ -481,10 +500,11 @@ object ZSink extends ZSinkPlatformSpecificConstructors {
     new ZSink(push) {}
 
   /**
-   * A sink that collects all of its inputs into a list.
+   * A sink that collects all of its inputs into a chunk.
    */
-  def collectAll[A]: ZSink[Any, Nothing, A, List[A]] =
-    foldLeftChunks(Chunk[A]())(_ ++ (_: Chunk[A])).map(_.toList)
+  def collectAll[A]: ZSink[Any, Nothing, A, Chunk[A]] =
+    foldLeftChunks[A, ChunkBuilder[A]](ChunkBuilder.make[A]())((b, chunk) => b ++= chunk)
```
Comment (adamgfraser) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala:506]: I believe the issue is that `ChunkBuilder.make()` is not wrapped in an effect constructor and is violating referential transparency.

**Comment by borissmidt** [comment]
Ah so i'm sharing a reference between a mutable array because its passed by
value and not as a thunk, that sounds logical to me. I'll fix it

**Comment by borissmidt** [comment]
I fixed, it but i'm not sure if it is a clean way of doing it, i tried t make a 'lazy chunkbuilder' that creates a new instance on the first += operation. But it wasn't possible with the typing of the return type.
An alternative could have been using Option. But then i wonder what the impact of the extra allocations would have been.

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 3]
```diff
@@ -481,10 +500,25 @@ object ZSink extends ZSinkPlatformSpecificConstructors {
...
   /**
-   * A sink that collects all of its inputs into a list.
+   * A sink that collects all of its inputs into a chunk.
    */
-  def collectAll[A]: ZSink[Any, Nothing, A, List[A]] =
-    foldLeftChunks(Chunk[A]())(_ ++ (_: Chunk[A])).map(_.toList)
+  def collectAll[A]: ZSink[Any, Nothing, A, Chunk[A]] =
+    //Cannot make a builder that initializes itself on the first operation.
+    // falling back to null instead of an option/None combination
+    foldLeftChunks[A, ChunkBuilder[A]](null: ChunkBuilder[A])((b, chunk) =>
```
Comment (adamgfraser) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala:508]: Can we just wrap the normal `ChunkBuilder.make()` in `ZIO.effectTotal`? That is going to be the easiest way to solve this?

Comment (borissmidt) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala:508]: Then you are stuck with the `Sink[Any, Nohting,A, UIO[Chunkbuilder[A]]]` where you cannot get the Chunkbuilder out of the UIO

Comment (iravid) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala:508]: @borissmidt
```scala
ZSink {
  for {
    builder <- UIO(ChunkBuilder.make()).toManaged_
    foldingSink = foldLeftChunks(...)
    push <- foldingSink.push
  } yield push
}
```

File: streams-tests/shared/src/test/scala/zio/stream/ZSinkSpec.scala [hunk 4]
```diff
@@ -25,6 +25,17 @@ object ZSinkSpec extends ZIOBaseSpec {
             .run(ZSink.collectAllToMap((_: Int) % 3)(_ + _))
         )(equalTo(Map[Int, Int](0 -> 18, 1 -> 12, 2 -> 15)))
       ),
+//      testM("collectAllToMap referential transparancy") {
```
Comment (iravid) [inline streams-tests/shared/src/test/scala/zio/stream/ZSinkSpec.scala:28]: Can we decide on the fate of this test? Either it stays uncommented or it gets deleted :-)

File: streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala [hunk 5]
```diff
@@ -160,21 +160,21 @@ object ZStreamSpec extends ZIOBaseSpec {
...
           },
           testM("with a sink that always signals more") {
             assertM(
               ZStream(1, 2, 3)
                 .aggregate(ZTransducer.fold(0)(_ => true)(_ + _))
                 .runCollect
-            )(equalTo(List(1 + 2 + 3)))
+            )(equalTo(Chunk(1 + 2 + 3)))
           },
           // testM("managed") {
```
Comment (iravid) [inline streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala:179]: Do you mind deleting the commented lines while you're here? 🙏🏻

**Comment by borissmidt** [comment]
I think it should be ok now? i've added the missing merge conflicts.

File: streams-tests/shared/src/test/scala/zio/stream/ZSinkSpec.scala (hunk 4 above)
Comment (borissmidt) [reply streams-tests/shared/src/test/scala/zio/stream/ZSinkSpec.scala]: I'm just not well versed in the zio test libary yet to easily implement it. I need more experience first.

**Comment by borissmidt** [comment]
I ran the fmt and fix command but there are no more changes detected. so i don't know why the fmtCheck still fails. Is there any reason why fix and fmt aren't ran during the build?

**Comment by iravid** [comment]
Use the `prepare` command in sbt. I think the contributor guide mentions that.

The CI doesn't commit to Github, so it doesn't format the code there.

**Comment by borissmidt** [comment]
Aha you have to do `sbt clean prepare` otherwise it has chached the formatting and will not do anything. I've more experience with gradle.

**Comment by borissmidt** [comment]
Oke i fixed all issues

**Comment by borissmidt** [comment]
@iravid Could this be merged before there is another merge conflict?

**Comment by iravid** [comment]
Thank you @borissmidt!

**Pull request merged**